### PR TITLE
Make api_compatibility_test pass in Python 3

### PR DESCRIPTION
**Pull request by girving**
I was (perhaps masochistically) curious what it would take.

This is a bunch of hacks, but it isn't that much overall code.  I leave it up to others whether it's worth ignoring, cleaning up, or adopting with minimal changes.  Main plus: nearly everyone outside Google uses Python 3, and this would make it easier for people to contribute.

### CODE REVIEW

**Review by annarev — DISMISSED** [review]

**Comment by girving** [comment]
To clarify: this doesn't remotely pass pylint, so it shouldn't be approved without some modifications.  Happy to make those modifications if we think syntax issues are the only problem, but I'd want explicit acknowledge that this sort of hackery is okay. :)

File: tensorflow/tools/api/lib/python_object_to_proto_visitor.py [hunk 1]
```diff
@@ -172,7 +229,7 @@ def _AddMember(member_name, member_obj, proto):
       elif tf_inspect.isclass(parent):
         # Construct a class.
         class_obj = api_objects_pb2.TFAPIClass()
-        class_obj.is_instance.extend(_SanitizedMRO(parent))
+        class_obj.is_instance.extend(_normalize_isinstance(i) for i in _SanitizedMRO(parent))
```
Comment (annarev) [inline tensorflow/tools/api/lib/python_object_to_proto_visitor.py]: Line too long.

File: tensorflow/tools/api/lib/python_object_to_proto_visitor.py [hunk 2]
```diff
@@ -33,6 +35,59 @@
...
+  _CODE_ATTR = '__code__'
+  _CLASS_TO_TYPE = {}
+  for t in ('property', 'object', 'getset_descriptor', 'int', 'str', 'type',
+            'tuple', 'module', 'collections.defaultdict', 'set', 'dict',
+            'NoneType', 'frozenset'):
+    _CLASS_TO_TYPE["<class '%s'>" % t] = "<type '%s'>" % t
+  for e in 'Exception', 'RuntimeError':
+    _CLASS_TO_TYPE["<class '%s'>" % e] = "<type 'exceptions.%s'>" % e
+  _NORMALIZE_ISINSTANCE = {
+      "<class 'tensorflow.python.training.monitored_session._MonitoredSession.StepContext'>":
```
Comment (annarev) [inline tensorflow/tools/api/lib/python_object_to_proto_visitor.py]: Add `# pylint: disable=line-too-long` at the end of this line

File: tensorflow/tools/api/lib/python_object_to_proto_visitor.py [hunk 3]
```diff
@@ -19,6 +19,8 @@
 from __future__ import division
 from __future__ import print_function
 
+import re
```
Comment (annarev) [inline tensorflow/tools/api/lib/python_object_to_proto_visitor.py]: remove unused import

**Comment by annarev** [comment]
@girving I left some comments including lint issues caught by "Ubuntu Sanity" build.
In addition to these comments, can you also remove `import unittest` in api_compatibility_test.py? It is no longer used with `@unittest.skipUnless` decorators removed.

Otherwise, this change is ok. Indeed having these manually-specified replacements doesn't look very nice. But I don't see a better way to have this test work with both Python 2 and Python 3.

File: tensorflow/tools/api/lib/python_object_to_proto_visitor.py [hunk 4]
```diff
@@ -33,6 +35,59 @@
...
+  def _normalize_type(ty):
+    return _CLASS_TO_TYPE.get(ty, ty)
+
+  def _normalize_isinstance(ty):
+    return _NORMALIZE_ISINSTANCE.get(ty, ty)
+
+  def _skip_member(cls, member):
+    if member == 'with_traceback':
+      return True
+    if (cls in ('VariableSynchronization', 'UnconnectedGradients', 'VariableAggregation') and
```
[the remark this answers is not in the document]
Comment (girving) [reply tensorflow/tools/api/lib/python_object_to_proto_visitor.py]: Done.

File: tensorflow/tools/api/lib/python_object_to_proto_visitor.py [hunk 5]
```diff
@@ -33,6 +35,59 @@
...
+  def normalize_proto(proto):
+    for kind in 'tf_module', 'tf_class':
+      if proto.HasField(kind):
+        for member in getattr(proto, kind).member:
+          if member.mtype == "<class 'abc.ABCMeta'>":
+            member.mtype = "<type 'type'>"
+        if proto.path == 'tensorflow.train.NanLossDuringTrainingError':
+          del proto.tf_class.member[1]
+else:
+  _CODE_ATTR = 'func_code'
```
[the remark this answers is not in the document]
Comment (girving) [reply tensorflow/tools/api/lib/python_object_to_proto_visitor.py]: Done.

File: tensorflow/tools/api/lib/python_object_to_proto_visitor.py [hunk 6]
```diff
@@ -33,6 +35,59 @@
...
+      return True
+
+  def normalize_proto(proto):
+    for kind in 'tf_module', 'tf_class':
+      if proto.HasField(kind):
+        for member in getattr(proto, kind).member:
+          if member.mtype == "<class 'abc.ABCMeta'>":
+            member.mtype = "<type 'type'>"
+        if proto.path == 'tensorflow.train.NanLossDuringTrainingError':
+          del proto.tf_class.member[1]
```
[the remark this answers is not in the document]
Comment (girving) [reply tensorflow/tools/api/lib/python_object_to_proto_visitor.py]: Done.

File: tensorflow/tools/api/lib/python_object_to_proto_visitor.py [hunk 7]
```diff
@@ -33,6 +35,59 @@
...
+      return True
+    if (cls in ('VariableSynchronization', 'UnconnectedGradients', 'VariableAggregation') and
+        member in ('name', 'value')):
+      return True
+
+  def normalize_proto(proto):
+    for kind in 'tf_module', 'tf_class':
+      if proto.HasField(kind):
+        for member in getattr(proto, kind).member:
+          if member.mtype == "<class 'abc.ABCMeta'>":
```
[the remark this answers is not in the document]
Comment (girving) [reply tensorflow/tools/api/lib/python_object_to_proto_visitor.py]: Done.

**Comment by girving** [comment]
Thanks for the comments!  Should be all resolved.  I agree that taking the hit once of updating the goldens once to get code simplicity (and symmetry) is good.

**Review by annarev — APPROVED** [review]

**Comment by annarev** [comment]
Thank you for fixing the test to work with python 3!

**Comment by girving** [comment]
You're welcome!

**Comment by martinwicke** [comment]
There's a couple of failures. I haven't look too deeply, is this just a concurrent change or is this a difference to Py2 (maybe preferring more local classes in superclass lists?):

```
  path: "tensorflow.train.MonitoredSession.StepContext"
  tf_class {
-   is_instance: "<class \'google3.third_party.tensorflow.python.training.monitored_session.StepContext\'>"
+   is_instance: "<class \'google3.third_party.tensorflow.python.training.monitored_session._MonitoredSession.StepContext\'>"
```

and 

```
  path: "tensorflow.train.SingularMonitoredSession.StepContext"
  tf_class {
-   is_instance: "<class \'google3.third_party.tensorflow.python.training.monitored_session.StepContext\'>"
+   is_instance: "<class \'google3.third_party.tensorflow.python.training.monitored_session._MonitoredSession.StepContext\'>"
```

and

```
  path: "tensorflow.Variable.SaveSliceInfo"
  tf_class {
-   is_instance: "<class \'google3.third_party.tensorflow.python.ops.variables.SaveSliceInfo\'>"
+   is_instance: "<class \'google3.third_party.tensorflow.python.ops.variables.Variable.SaveSliceInfo\'>"
```

**Comment by girving** [comment]
@martinwicke There's specific logic to fix those differences, which is why the tests pass in both Python 2 and Python 3: https://github.com/tensorflow/tensorflow/pull/23006/files#diff-66076c6ed184cb1cc7997df09d0b1c6fR50.  I'm not sure why the copybara test introduces new failures.

**Pull request merged**